build/gen_stub.php

```diff
@@ -1262,29 +1264,89 @@ function initPhpParser() {
-$context->forceRegeneration =
-    isset($options["f"]) || isset($options["force-regeneration"]) || $printParameterStats;
+$verify = isset($options["verify"]);
+$context->forceRegeneration = isset($options["f"]) || isset($options["force-regeneration"]);
+$context->forceParse = $context->forceRegeneration || $printParameterStats || $verify;
 
 if (isset($options["h"]) || isset($options["help"])) {
-    die("\nusage: gen-stub.php [ -f | --force-regeneration ] [ --parameter-stats ] [ -h | --help ] [ name.stub.php | directory ]\n\n");
+    die("\nusage: gen-stub.php [ -f | --force-regeneration ] [ --parameter-stats ] [ --verify ] [ -h | --help ] [ name.stub.php | directory ]\n\n");
 }
 
+$fileInfos = [];
 $location = $argv[$optind] ?? ".";
 if (is_file($location)) {
     // Generate single file.
-    processStubFile($location, $context);
+    $fileInfo = processStubFile($location, $context);
+    if ($fileInfo) {
+        $fileInfos[] = $fileInfo;
+    }
 } else if (is_dir($location)) {
-    processDirectory($location, $context);
+    $fileInfos = processDirectory($location, $context);
 } else {
     echo "$location is neither a file nor a directory.\n";
     exit(1);
 }
 
 if ($printParameterStats) {
-    arsort($context->parameterStats);
-    echo json_encode($context->parameterStats, JSON_PRETTY_PRINT), "\n";
+    $parameterStats = [];
+
+    foreach ($fileInfos as $fileInfo) {
+        foreach ($fileInfo->getAllFuncInfos() as $funcInfo) {
+            foreach ($funcInfo->args as $argInfo) {
+                if (!isset($context->parameterStats[$argInfo->name])) {
+                    $parameterStats[$argInfo->name] = 0;
+                }
+                $parameterStats[$argInfo->name]++;
+            }
+        }
+    }
+
+    arsort($parameterStats);
+    echo json_encode($parameterStats, JSON_PRETTY_PRINT), "\n";
+}
+
+if ($verify) {
+    $errors = [];
+    $funcMap = [];
+    $aliases = [];
+
+    foreach ($fileInfos as $fileInfo) {
+        foreach ($fileInfo->getAllFuncInfos() as $funcInfo) {
+            /** @var FuncInfo $funcInfo */
+            $funcMap[$funcInfo->name->__toString()] = $funcInfo;
+
+            if ($funcInfo->aliasType === "alias") {
+                $aliases[] = $funcInfo;
+            }
+        }
+    }
+
+    foreach ($aliases as $alias) {
+        if ($alias->alias === null || !isset($funcMap[$alias->alias->__toString()])) {
+            $errors[] = "Aliased function {$alias->alias}() cannot be found";
+            continue;
+        }
+
+        $aliasedFunc = $funcMap[$alias->alias->__toString()];
+
+        $aliasedArgMap = [];
+        foreach ($aliasedFunc->args as $arg) {
+            $aliasedArgMap[$arg->name] = $arg;
+        }
+
+        foreach ($alias->args as $arg) {
+            if (!isset($aliasedArgMap[$arg->name])) {
+                $errors[] = "{$alias->name}(): Argument \$$arg->name of aliased function {$aliasedFunc->name}() is missing";
+            }
+        }
+    }
```

Comment:
How/why does this work with method<->function aliases? Won't they fail this check due to the first param?

Reply:
No, they won't, because I'm checking the *alias* function's parameters *against the aliased* function's parameters, not the opposite way. So this solution assumes that method's are the aliases... Which is maybe not that nice to do for external extensions which we can't control.

If the solution I proposed feels hacky for you, I'm ok to just merge the actual fixes, and leave gen_stub.php alone.

Reply:
I do think we should include this in gen_stub (and the implementation actually turned out to be a lot simpler than I expected), but we should probably handle the method/function case right. I think it should not be too hard to do this using something along these lines:

```
$args = $func->args;
$aliasedArgs = $aliasedFunc->args;
if ($func->isInstanceMethod() != $aliasedFunc->isInstanceMethod()) {
    if ($func->isInstanceMethod()) {
        $args = array_slice($args, 1);
    }
    if ($aliasedFunc->isInstanceMethod()) {
        $aliasArgs = array_slice($aliasedArgs, 1);
    }
}

// Compare $args and $aliasedArgs in lock-step.
```

Basically, if we're aliasing between a function and a non-static method, then we need to ignore the first argument on the method.

Reply:
Aha! I like the idea very much. :)

P.S. I'm not exactly sure what lock-step means 😅  but if it is a synonym to vica-versa, then yes, that's what I'll do :)

Reply:
Lock-step = zip :) In PHP we can use array_map for this:

```
array_map(function(?Param $arg, ?Param $aliasedArg) {
    // Compare...
}, $args, $aliasedArgs)
```

Reply:
So.. I managed to verify in lock steps :) I even added validation for the type and default value... However there's an issue with methods that are non-static, but parse parameters as it they were static... For now, I marked them as static, but I doubt that we can do this now, (right?). The other solution we could do is to mark them as `@method-alias`es, so that we find a use-case for the original idea (or should we rather call them as `@static-method-alias`?).

Reply:
Yeah, even the tests fail after the conversion to static methods, so we really can't choose this approach.